A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,296 @@
+
+
+## Background
+
+Some users of proxyless gRPC need to use "stateful session affinity" in gRPC.
+In this usage scenario, backend services maintain local state for RPC sessions
+and once a particular "session" is assigned to a particular backend (by the
+load-balancer), all future RPCs of that session must be routed to the same
+backend.
+
+This can be achieved by using HTTP cookies in the following manner.
+A gRPC client application sends the first RPC in a session without
+a cookie in the RPC. gRPC routes that RPC to some backend B1 based on the
+configured LB policy. When gRPC receives the response for that first RPC from
+B1, it attaches a configured cookie to the response before delivering the
+response to the client application. The cookie encodes some identifying
+information about B1 (for example its IP-address). The client application
+saves the cookie and uses that cookie in all future RPCs of that session. gRPC
+uses the cookie and gets the encoded backend information to route the RPC to
+the same backend B1 if it is still in a valid state.
+
+### Related Proposals: 
+
+* [gRFC A27: xDS-Based Global Load Balancing][A27]
+* [A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+This design involves the following 4 parts.
+
+### Listener and Route Configuration through xDS
+
+The feature is enabled and configured using the
+[cookie based stateful session extension][cookie-ext] which
+is an [http-filter][] in the HttpConnectionManager for the target service. A
+sample configuration (expressed as yaml) is as follows:
+
+```yaml
+   http_filters:
+   - name: envoy.filters.http.stateful_session
+     typed_config:
+       "@type": type.googleapis.com/envoy.extensions.filters.http.stateful_session.v3.StatefulSession
+       session_state:
+         name: envoy.http.stateful_session.cookie
+         typed_config:
+           "@type": type.googleapis.com/envoy.extensions.http.stateful_session.cookie.v3.CookieBasedSessionState
+           cookie:
+             name: global-session-cookie
+             path: /Package1.Service2/Method3
+             ttl: 120s
+```
+
+`name` is the name of the cookie used in this feature.
+
+`path` is the path for which this cookie is applied.
+
+`ttl` is the time-to-live for the cookie. It is set on the cookie but will not
+be enforced by gRPC.
+
+This new filter will be added to the HTTP Filter Registry and processed as
+described in [A39: xDS HTTP Filter Support][A39]. Specifically when this filter
+is present in the configuration, gRPC will create the appropriate client filter
+(aka interceptor in Java/Go) and install it in the channel to process data
+plane RPCs. We call this filter `CookieBasedStatefulSessionFilter`. It will
+copy the 3 configuration values - `name`, `ttl`  and `path` -  into the filter
+object. Validation and processing are further described in
+[Stateful session][stateful-session]. Note that
+[StatefulSessionPerRoute][stateful-session-per-route] *will* be supported.
```

Comment:
There seems to be no reference to that in the detailed description below.

Reply:
Good point - I will add the spec to the detailed description.

Reply:
Done, PTAL